### PR TITLE
(feat) O3-3192: Harvest bed administration functionality from ugandaemr into patientmanagement

### PR DESCRIPTION
**Pull request by brandones**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

Copied the code from https://github.com/METS-Programme/esm-ugandaemr-bed-management-app to create a bed management app here. Code required no modification, other than to fix some imports. The `ErrorState` component now comes from esm-framework rather than esm-patient-common-lib, and the `CardHeader` component has been copied into this repo.

Since the code is all copy-pasted from https://github.com/METS-Programme/esm-ugandaemr-bed-management-app, I would not as a reviewer bother to review most of the files. If we want to improve things we should do that in follow-up PRs. Please do check the boilerplate stuff such as `package.json`.

## Screenshots

Seems to work. Here is bed management running from this PR.

https://github.com/user-attachments/assets/f080e7a7-06f8-4e79-ab4a-dd3e095c523e

## Related Issue

https://openmrs.atlassian.net/browse/O3-3192

### CODE REVIEW

**Comment by vasharma05** [comment]
Hi @brandones!
Thanks for the PR!
Can you also verify if the configuration defaults are set to work with dev3?
Thanks!

File: packages/esm-bed-management-app/src/config-schema.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { Type } from "@openmrs/esm-framework";
+
+export const configSchema = {
+  admissionLocationTagUuid: {
+    _type: Type.UUID,
+    _description:
+      "UUID for the location tag of the `Admission Location`. Patients may only be admitted to inpatient care in a location with this tag",
+    _default: "1c783dca-fd54-4ea8-a0fc-2875374e9cb6",
```
Comment (vasharma05) [inline packages/esm-bed-management-app/src/config-schema.ts:8]: No location tag `Admission Location` was found on the dev3
@ibacher, do we need to add a new tag in the Distro RefApp?
Thanks!

**Review by mseaton — CHANGES_REQUESTED** [review]
@brandones - see a few comments.  It looks to me like the entire esm has been harvested over - not just the files related to bed administration, but also those related to patient-bed assignments, which should not be included.

File: packages/esm-bed-management-app/src/config-schema.ts (hunk 1 above)
Comment (mseaton) [reply packages/esm-bed-management-app/src/config-schema.ts:8]: It would be best if we could broaden this to support a `uuid or name` of a location tag, and then configure this by default to be `Admission Location`, which is what is expected throughout the backend modules, including EMR API and Bed Management.  I do not support hard-coding uuids as defaults in esm configs in general.

File: packages/esm-bed-management-app/src/assets/landing-page.png
Comment (ibacher) [inline packages/esm-bed-management-app/src/assets/landing-page.png:1]: I'm also not sure why we need this in the package?

File: packages/esm-bed-management-app/src/summary/summary.resource.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  return {
+    data: data?.data?.results ?? [],
+    error,
+    isLoading,
+    isValidating,
+    mutate,
+  };
+};
+
+export const getBedsForLocation = (locationUuid: string) => {
```
Comment (ibacher) [inline packages/esm-bed-management-app/src/summary/summary.resource.ts:22]: I sort of understand having a `use` hook and a `get` function for the same thing (though, not really), but `getBedsForLocation` and `findBedByLocation` are basically duplicative.

File: packages/esm-bed-management-app/src/summary/summary.resource.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+  return {
+    bedData: mappedBedData,
+    isLoading,
+    error,
+  };
+};
+
+export const useLocationName = (locationUuid: string) => {
+  const apiUrl = `/ws/rest/v1/location/${locationUuid}`;
```
Comment (ibacher) [inline packages/esm-bed-management-app/src/summary/summary.resource.ts:55]: This feels like a case where we should've already been able to get this data without making a new request.

```suggestion
  const apiUrl = `/ws/rest/v1/location/${locationUuid}?v=custom:(display)`;
```

File: packages/esm-bed-management-app/src/summary/summary.resource.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+};
+
+export const findBedByLocation = (locationUuid: string) => {
+  const locationsUrl = `/ws/rest/v1/bed?locationUuid=${locationUuid}`;
+  return openmrsFetch(locationsUrl, {
+    method: 'GET',
+  });
+};
+
+export const useWards = (locationUuid: string) => {
```
Comment (ibacher) [inline packages/esm-bed-management-app/src/summary/summary.resource.ts:72]:
```suggestion
export const useWards = (tag: string) => {
```

File: packages/esm-bed-management-app/src/summary/summary.resource.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+};
+interface BedType {
+  name: string;
+  displayName: string;
+  description: string;
+}
+interface BedTag {
+  name: string;
+}
+export async function saveBedType({ bedPayload }): Promise<FetchResponse<BedType>> {
```
Comment (ibacher) [inline packages/esm-bed-management-app/src/summary/summary.resource.ts:153]: The parameter accepted here should be typed.

File: packages/esm-bed-management-app/src/summary/summary.scss [hunk 6]
```diff
@@ -0,0 +1,72 @@
+@use '@carbon/colors';
+@use '@carbon/layout';
+
+.cardContainer {
```
Comment (ibacher) [inline packages/esm-bed-management-app/src/summary/summary.scss:4]: Formatting here should be consistent. Did we run prettier on this at all?

**Review by ibacher — COMMENTED** [review]
Most comments I made are placeholders for things that should be fixed, so I'll surface the two that are relevant:

1. I don't see why we need the PNG file at all
2. At least one of the SCSS files has very inconsistent spacing, which should be fixed.

File: packages/esm-bed-management-app/src/config-schema.ts (hunk 1 above)
Comment (mogoodrich) [reply packages/esm-bed-management-app/src/config-schema.ts:8]: Agreed, this property should be something like "admission location tag" instead of "admission location uuid" and default to "Admission Location"

Comment (brandones) [reply packages/esm-bed-management-app/src/config-schema.ts:8]: Yeah I would love it if we could reference things by name in the configs, but I'm not aware of the location endpoint will query locations by location tag name. If that is possible, it's not documented in the [REST docs](https://rest.openmrs.org/#location). Though it should be noted that neither is getting locations by location tag UUID documented there. Not sure if there's some other source of documentation or people just read the backend Java code to figure out what the API is.

The other possibility is to support using location tag name in the config by making two requests. First we would get all the location tag data from `/locationtag`, then look up the UUID by the configured name, then use that UUID in the `/location` query.

Comment (ibacher) [reply packages/esm-bed-management-app/src/config-schema.ts:8]: Well, the REST docs are incomplete on that point; we [support both](https://github.com/openmrs/openmrs-module-webservices.rest/blob/57b2e5ed9ce59201a351093ea68e62de408799ff/omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/LocationSearchHandler.java#L62).

Comment (brandones) [reply packages/esm-bed-management-app/src/config-schema.ts:8]: Oh I replaced the UUID with a tag name and it just works :laughing:  We should probably document this somewhere

**Comment by brandones** [comment]
@ibacher @mseaton What's the deal with the `/admissionLocation` REST endpoint? Presumably this comes from [module-bedmanagement](https://github.com/openmrs/openmrs-module-bedmanagement)? Is it documented anywhere?

Right now there is a function called `useAdmissionLocations` that makes a request to that endpoint. Semantically I don't know why that should be a different thing than making a request to the `/location` endpoint with the "Admission Location" tag.

File: packages/esm-bed-management-app/src/summary/summary.resource.ts (hunk 2 above)
Comment (brandones) [reply packages/esm-bed-management-app/src/summary/summary.resource.ts]: `getBedsForLocation` is cruft. `findBedByLocation` is called many times by the [same component](https://github.com/openmrs/openmrs-esm-patient-management/blob/74a79e963fa561a900d3cc0d6616760e09a6bd33/packages/esm-bed-management-app/src/bed-administration/bed-administration-table.component.tsx#L87-L108) to fetch beds for an array of locations. This can probably be simplified (or all the logic moved into a hook, at least).

**Comment by ibacher** [comment]
@brandones Yeah, it's poorly named class. `AdmissionLocation` essentially provides bed-management data about a location, e.g., the number of beds and the number of beds occupied. See [this class](https://github.com/openmrs/openmrs-module-bedmanagement/blob/master/api/src/main/java/org/openmrs/module/bedmanagement/AdmissionLocation.java). APIs for modules are generally much less documented than the APIs for REST, and APIs for modules that have largely not been a part of the RefApp prior to O3 are usually almost completely undocumented.

File: packages/esm-bed-management-app/src/summary/summary.resource.ts (hunk 2 above)
Comment (ibacher) [reply packages/esm-bed-management-app/src/summary/summary.resource.ts]: That's definitely something that should go in the "improve the API" queue.

File: packages/esm-bed-management-app/src/config-schema.ts (hunk 1 above)
Comment (mseaton) [reply packages/esm-bed-management-app/src/config-schema.ts]: TBH, in the future my hope is that we don't need this Admission Location tag at all.  I don't really see why the bedmanagement module should restrict which locations beds can be assigned to to only "Admission Location" locations.  There really is no need for that, and in fact it may be that we want to support assigning patients to beds without actually admitting them, as was indicated in a recent slack conversation...  But that's for a future ticket.

**Comment by mseaton** [comment]
> Yeah, it's poorly named class

+1.  It may make sense to deprecate and rename this more sensibly in the future.

File: packages/esm-bed-management-app/src/config-schema.ts (hunk 1 above)
Comment (mseaton) [reply packages/esm-bed-management-app/src/config-schema.ts]: > If we remove the tag, how will we distinguish locations which are not wards? We still want a way to mark a location as admission location since not every location can admit patients

I'm not saying we should remove the tag from the back end, or from use in appropriate places.  But all we are aiming to harvest here is a UI for creating beds and assigning beds to a location, as well as managing bed types and tags.  Whether or not the locations we are assigning beds to are considered inpatient wards that can admit patients is not necessarily a factor.

Comment (mseaton) [reply packages/esm-bed-management-app/src/config-schema.ts]: For now, I think Admission Location is used on the backend as well, so this should stay as-is.  I was just commenting on the fact that this might go away in the future.  A new location tag would be fine if we need to distinguish.  Personally, I see no harm in an admin UI for adding beds to locations broadly.  Then, the locations that users are allowed to assign patients to beds are those locations that have beds allocated to them...

File: packages/esm-bed-management-app/src/bed-administration/bed-administration-form.component.tsx [hunk 7]
```diff
@@ -93,7 +93,7 @@ const BedAdministrationForm: React.FC<BedAdministrationFormProps> = ({
     resolver: zodResolver(BedAdministrationSchema),
     defaultValues: {
       bedId: initialData.bedNumber || '',
-      description: initialData.description || '',
+      description: initialData.bedType.description || '',
```
Comment (mseaton) [inline packages/esm-bed-management-app/src/bed-administration/bed-administration-form.component.tsx:96]: Is this really intended to be the bedType.description?  BedFormData does appear to have a description property.  Why was this changed?

File: packages/esm-bed-management-app/src/bed-admission/bed-layout/bed-layout.component.tsx
Comment (mseaton) [inline packages/esm-bed-management-app/src/bed-admission/bed-layout/bed-layout.component.tsx:1]: Are you sure these bed-layout components should be deleted?  I can't see here what they were doing, but if these are responsible for creating a layout of beds within a location then that is part of bed administration we need to keep.

**Review by mseaton — COMMENTED** [review]
A bit hard to review.  Looks like lots of good changes.  I made a few comments on a few things to check on.

File: packages/esm-bed-management-app/src/bed-administration/bed-administration-form.component.tsx (hunk 7 above)
Comment (brandones) [reply packages/esm-bed-management-app/src/bed-administration/bed-administration-form.component.tsx:96]: The `description` is an attribute of the bed type on the backend. AFAIK beds themselves do not have descriptions. The form data has a description field, which I think is why it is represented at the top level of that object.

File: packages/esm-bed-management-app/src/bed-admission/bed-layout/bed-layout.component.tsx
Comment (brandones) [reply packages/esm-bed-management-app/src/bed-admission/bed-layout/bed-layout.component.tsx:1]: No, these are responsible for displaying the layout of beds when allocating patients to beds. See [<BedLayoutList>](https://github.com/METS-Programme/esm-ugandaemr-bed-management-app/blob/e094746e7ef12d86b885a0dca9b955be2b84265e/src/bed-admission/bed-layout/bed-layout-list.component.tsx#L72) which is only used in [`allocate-bed-workspace.component.tsx`](https://github.com/METS-Programme/esm-ugandaemr-bed-management-app/blob/main/src/workspace/allocate-bed-workspace.component.tsx#L106).

**Comment by brandones** [comment]
Please excuse the force-push, wiping out previous comments. But I've made this much easier to review (hopefully) on a per-commit basis. Now all of the changes I've made are in the commit https://github.com/openmrs/openmrs-esm-patient-management/pull/1230/commits/db036dc7869d7d66e40c40102e002cf79b2bdd03 . The prior commits are just copying the files in and then prettier & lint fix.

**Review by mseaton — APPROVED** [review]
Great work @brandones .  LGTM, though I only gave it a rather cursory review this time around.  Would be great to get someone from the Uganda team to look it over and confirm they'd be comfortable using this shared code for bed administration.

**Comment by brandones** [comment]
Tagging contributors to the https://github.com/METS-Programme/esm-ugandaemr-bed-management-app repo: @denniskigen @slubwama @gabriel090
